### PR TITLE
Update code that interacts with Google Analytics to use sendBea…

### LINKED ISSUE #8605: Update code that interacts with Google Analytics to use sendBeacon
The script that we use for interacting with Google Analytics (GA) has an [option](https://developers.google.com/analytics/devguides/collection/analyticsjs/sending-hits#specifying_different_transport_mechanisms) that allows us to specify we want it to use [`sendBeacon`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon) as its transport mechanism. 

Using `sendBeacon` will minimize the number of events that get lost due to a page unloading while we are waiting for the event to be sent.

This issue is about updating the code in `core/tracking.js` to use this new option.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #8605

### CODE REVIEW

**Comment by bobsilverberg** [comment]
> It'd be nice to update the CSP config as part of this patch. I am not sure but maybe we simply need to remove this line for AMO:
> 
> https://github.com/mozilla/addons-frontend/blob/b6674253464ddfc7e7c4e4c4f57274ba631fa3ef/config/default-amo.js#L25
> , and a similar one for the disco pane.

I actually hadn't planned on updating the code for disco pane, as my understanding is that it's pretty much not used anymore (except by older clients, who might not even have support for `sendBeacon`). Do you think it's important to update disco for this as well?

**Comment by bobsilverberg** [comment]
> 
> I actually hadn't planned on updating the code for disco pane, as my understanding is that it's pretty much not used anymore (except by older clients, who might not even have support for `sendBeacon`). Do you think it's important to update disco for this as well?

Oops, my mistake. There isn't any code specific to `disco` for this. I noticed that there was a file called `disco/tracking.js` so I assumed that was related to GA, but it isn't.

**Comment by willdurand** [comment]
> Do you think it's important to update disco for this as well?

Yeah, we're still maintaining it and we use GA there too: https://github.com/mozilla/addons-frontend/blob/b3eff6926b8b2b94eb5b09cd489f464504439406/config/default-disco.js#L58-L59

**Comment by bobsilverberg** [comment]
Thanks @willdurand. I've updated the CSP in the config files as well. I tested locally and it all seems to work. I do see this message in the console:

`Content Security Policy: The page’s settings observed the loading of a resource at https://www.google-analytics.com/collect (“connect-src”). A CSP report is being sent.`

but it doesn't seem to interfere with the beacon being sent (I can see it in the network panel).

**Comment by muffinresearch** [comment]
> Thanks @willdurand. I've updated the CSP in the config files as well. I tested locally and it all seems to work. I do see this message in the console:
> 
> `Content Security Policy: The page’s settings observed the loading of a resource at https://www.google-analytics.com/collect (“connect-src”). A CSP report is being sent.`
> 
> but it doesn't seem to interfere with the beacon being sent (I can see it in the network panel).

CSP isn't enforced locally IIRC. See this comment details of the configuration needed: https://github.com/mozilla/addons-frontend/issues/8605#issuecomment-532264117

**Comment by bobsilverberg** [comment]
Thanks @muffinresearch, I'm adding the entry for `connect-src` now.

**Comment by bobsilverberg** [comment]
@willdurand I added the GA host to `connectSrc`. This should be ready to review now.

File: config/default-amo.js [hunk 1]
```diff
@@ -22,7 +22,6 @@ module.exports = {
         'data:',
         amoProdCDN,
         staticHost,
-        'https://www.google-analytics.com',
```
Comment (willdurand) [inline config/default-amo.js]: On Line 17, there is the same URL suffixed with the js file. I guess we could reuse `analyticsHost` for this value too.

File: config/default-disco.js [hunk 2]
```diff
@@ -55,7 +55,6 @@ module.exports = {
         'data:',
         amoProdCDN,
         staticHost,
-        'https://www.google-analytics.com',
```
Comment (willdurand) [inline config/default-disco.js]: On Line 50, there is the same URL suffixed with the js file. I guess we could reuse `analyticsHost` for this value too.

File: config/dev-amo.js [hunk 3]
```diff
@@ -22,7 +22,6 @@ module.exports = {
         'data:',
         amoDevCDN,
         staticHost,
```
Comment (willdurand) [inline config/dev-amo.js]: Please replace the google analytics host with the variable you introduced :)

File: config/dev-disco.js [hunk 4]
```diff
@@ -17,7 +17,6 @@ module.exports = {
         'data:',
         amoDevCDN,
         staticHost,
-        'https://www.google-analytics.com',
```
Comment (willdurand) [inline config/dev-disco.js]: `config/uitests-disco.js` has not been updated I believe.

**Comment by bobsilverberg** [comment]
> r+wc
> 
> It'd be nice to update `config/development.js` (or maybe `config/development-amo.js`) to get rid of the "report only" CSP violations. This would mean applying pretty much the same changes.

Thanks @willdurand. I updated `config/development.js` and it made the error for script-src go away for me.

**Pull request merged**